File: DSA/Arshdeep-Kaur/Day 65/Day65_Question2.cpp
class Solution {
public:
    bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
        stack<int> temp;
        int n = pushed.size();
        int i=0, j=0;

        while(i < n && j < n) {
            temp.push(pushed[i]);
            while(!temp.empty() && j<n && popped[j] == temp.top()) {
                temp.pop();
                j++;
            }
            i++;
        }

        return temp.empty();
    }
};

File: DSA/Manav-Rajput/Day65_question2.java
class Solution {
    public boolean validateStackSequences(int[] pushed, int[] popped) {
        Stack<Integer> stack = new Stack<>();
        int popIndex = 0;
        
        for (int num : pushed) {
            stack.push(num); // Push the current element onto the stack
            
            // Check if the top of the stack matches the next element in the popped array
            while (!stack.isEmpty() && stack.peek() == popped[popIndex]) {
                stack.pop(); // Pop the element from the stack
                popIndex++;  // Move to the next element in the popped array
            }
        }
        
        // If all elements have been successfully popped, the sequence is valid
        return stack.isEmpty();
    }
}
